File: Midterm Lab Work 3/MP11FrequencyCount.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Scanner;

/*
 * Student Name: CANTAROS, RICH ANDREI
 * Machine Problem: MP11 - Frequency count for column values
 * Language: Java
 *
 * Description:
 * This program asks the user for the CSV dataset file path first.
 * It then reads and parses the CSV file using BufferedReader and FileReader.
 * After loading the dataset, the program asks the user which column to analyze.
 * It counts how many times each value appears in that column and displays the result.
 */
public class MP11FrequencyCount {

    /*
     * CSVData class:
     * - headers stores the CSV header row
     * - rows stores all valid data rows
     * - sourceLineNumbers stores the original line numbers from the file
     */
    static class CSVData {
        ArrayList<String> headers = new ArrayList<>();
        ArrayList<String[]> rows = new ArrayList<>();
        ArrayList<Integer> sourceLineNumbers = new ArrayList<>();
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.println("Student: CANTAROS, RICH ANDREI");
        System.out.println("MP11 - Frequency Count for Column Values");
        System.out.print("Enter CSV dataset file path: ");
        String filePath = scanner.nextLine();

        try {
            CSVData data = loadCSV(filePath);

            if (data.rows.isEmpty()) {
                System.out.println("No valid data rows were found in the dataset.");
                return;
            }

            System.out.println("\nAvailable columns:");
            for (int i = 0; i < data.headers.size(); i++) {
                String header = data.headers.get(i).trim();
                if (!header.isEmpty()) {
                    System.out.println((i + 1) + " - " + header);
                }
            }

            System.out.print("\nEnter column name or column number for frequency count: ");
            String columnInput = scanner.nextLine();

            int columnIndex = findColumnIndex(data.headers, columnInput);

            if (columnIndex == -1) {
                System.out.println("Invalid column. Please run the program again and enter a valid column name or number.");
                return;
            }

            // frequencyMap stores each unique value and its count
            HashMap<String, Integer> frequencyMap = new HashMap<>();

            // Loop through each row and count values in the selected column
            for (String[] row : data.rows) {
                String value = safeGet(row, columnIndex).trim();

                if (value.isEmpty()) {
                    value = "(EMPTY)";
                }

                frequencyMap.put(value, frequencyMap.getOrDefault(value, 0) + 1);
            }

            // Convert map to list so it can be sorted by frequency descending
            ArrayList<Map.Entry<String, Integer>> sortedEntries = new ArrayList<>(frequencyMap.entrySet());
            sortedEntries.sort(new Comparator<Map.Entry<String, Integer>>() {
                @Override
                public int compare(Map.Entry<String, Integer> a, Map.Entry<String, Integer> b) {
                    int countCompare = Integer.compare(b.getValue(), a.getValue());
                    if (countCompare != 0) {
                        return countCompare;
                    }
                    return a.getKey().compareToIgnoreCase(b.getKey());
                }
            });

            System.out.println("\n========== FREQUENCY COUNT ==========");
            System.out.println("Selected Column: " + data.headers.get(columnIndex));
            System.out.println("Total Valid Rows: " + data.rows.size());
            System.out.println("Unique Values: " + sortedEntries.size());
            System.out.println("=====================================");

            for (Map.Entry<String, Integer> entry : sortedEntries) {
                System.out.printf("%-45s : %d%n", entry.getKey(), entry.getValue());
            }

        } catch (IOException e) {
            System.out.println("Error: " + e.getMessage());
        } catch (Exception e) {
            System.out.println("Unexpected error: " + e.getMessage());
        }
    }

    /*
     * loadCSV function:
     * - Reads the CSV file
     * - Finds the actual header row
     * - Skips extra text above the real dataset
     * - Stores valid rows only
     */
    public static CSVData loadCSV(String filePath) throws IOException {
        CSVData data = new CSVData();
        BufferedReader reader = new BufferedReader(new FileReader(filePath));

        String line;
        boolean headerFound = false;
        int fileLineNumber = 0;
        int headerSize = 0;

        while ((line = reader.readLine()) != null) {
            fileLineNumber++;

            // Remove possible BOM character
            line = line.replace("\uFEFF", "");

            List<String> parsedLine = parseCSVLine(line);

            if (!headerFound) {
                if (isRealHeader(parsedLine)) {
                    data.headers.addAll(parsedLine);
                    headerFound = true;
                    headerSize = parsedLine.size();
                }
                continue;
            }

            while (parsedLine.size() < headerSize) {
                parsedLine.add("");
            }

            if (isRowEmpty(parsedLine)) {
                continue;
            }

            data.rows.add(parsedLine.toArray(new String[0]));
            data.sourceLineNumbers.add(fileLineNumber);
        }

        reader.close();

        if (!headerFound) {
            throw new IOException("CSV header row was not found.");
        }

        return data;
    }

    /*
     * parseCSVLine function:
     * - Parses one CSV line correctly
     * - Handles commas inside quoted text like "Lastname,Firstname"
     */
    public static List<String> parseCSVLine(String line) {
        ArrayList<String> fields = new ArrayList<>();
        StringBuilder currentField = new StringBuilder();
        boolean inQuotes = false;

        for (int i = 0; i < line.length(); i++) {
            char ch = line.charAt(i);

            if (ch == '"') {
                if (inQuotes && i + 1 < line.length() && line.charAt(i + 1) == '"') {
                    currentField.append('"');
                    i++;
                } else {
                    inQuotes = !inQuotes;
                }
            } else if (ch == ',' && !inQuotes) {
                fields.add(currentField.toString());
                currentField.setLength(0);
            } else {
                currentField.append(ch);
            }
        }

        fields.add(currentField.toString());
        return fields;
    }

    /*
     * isRealHeader function:
     * - Checks whether the current line is the real CSV header
     */
    public static boolean isRealHeader(List<String> fields) {
        if (fields.size() < 8) {
            return false;
        }

        return fields.get(0).trim().equalsIgnoreCase("Candidate")
                && fields.get(1).trim().equalsIgnoreCase("Student/ Faculty/ NTE")
                && fields.get(3).trim().equalsIgnoreCase("Exam")
                && fields.get(6).trim().equalsIgnoreCase("Score");
    }

    /*
     * isRowEmpty function:
     * - Returns true if all values in the row are empty
     */
    public static boolean isRowEmpty(List<String> row) {
        for (String field : row) {
            if (!field.trim().isEmpty()) {
                return false;
            }
        }
        return true;
    }

    /*
     * safeGet function:
     * - Safely gets a value from a row without causing an index error
     */
    public static String safeGet(String[] row, int index) {
        if (index >= 0 && index < row.length) {
            return row[index];
        }
        return "";
    }

    /*
     * findColumnIndex function:
     * - Accepts either a column number or a column name
     */
    public static int findColumnIndex(ArrayList<String> headers, String input) {
        input = input.trim();

        try {
            int number = Integer.parseInt(input);
            if (number >= 1 && number <= headers.size()) {
                return number - 1;
            }
        } catch (NumberFormatException e) {
            // Not a number, continue checking as text
        }

        for (int i = 0; i < headers.size(); i++) {
            if (headers.get(i).trim().equalsIgnoreCase(input)) {
                return i;
            }
        }

        return -1;
    }
}

File: Midterm Lab Work 3/mp11_frequency_count.js
/*
 * Student Name: CANTAROS, RICH ANDREI
 * Machine Problem: MP11 - Frequency count for column values
 * Language: JavaScript (Node.js)
 *
 * Description:
 * This program asks the user for the CSV dataset file path first.
 * It reads and parses the CSV file using the Node.js File System module.
 * After loading the dataset, it asks which column to analyze.
 * It then counts the frequency of values in that column and displays the result.
 */

const fs = require("fs");
const readline = require("readline");

// readline interface for user input
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

// askQuestion function:
// used to prompt the user and return the answer
function askQuestion(question) {
    return new Promise(resolve => rl.question(question, resolve));
}

// parseCSVLine function:
// parses one CSV line while handling commas inside quotes
function parseCSVLine(line) {
    const fields = [];
    let currentField = "";
    let inQuotes = false;

    for (let i = 0; i < line.length; i++) {
        const ch = line[i];

        if (ch === '"') {
            if (inQuotes && i + 1 < line.length && line[i + 1] === '"') {
                currentField += '"';
                i++;
            } else {
                inQuotes = !inQuotes;
            }
        } else if (ch === "," && !inQuotes) {
            fields.push(currentField);
            currentField = "";
        } else {
            currentField += ch;
        }
    }

    fields.push(currentField);
    return fields;
}

// checks if the current row is the real CSV header
function isRealHeader(fields) {
    if (fields.length < 8) return false;

    return fields[0].trim().toLowerCase() === "candidate" &&
           fields[1].trim().toLowerCase() === "student/ faculty/ nte" &&
           fields[3].trim().toLowerCase() === "exam" &&
           fields[6].trim().toLowerCase() === "score";
}

// checks if the row is fully empty
function isRowEmpty(row) {
    return row.every(field => field.trim() === "");
}

// safely gets a value from a row
function safeGet(row, index) {
    return index >= 0 && index < row.length ? row[index] : "";
}

// loadCSV function:
// reads the file, finds the actual header row, and stores valid data rows
function loadCSV(filePath) {
    const content = fs.readFileSync(filePath, "utf8").replace(/\uFEFF/g, "");
    const lines = content.split(/\r?\n/);

    const data = {
        headers: [],
        rows: [],
        sourceLineNumbers: []
    };

    let headerFound = false;
    let headerSize = 0;

    lines.forEach((line, lineIndex) => {
        const parsedLine = parseCSVLine(line);

        if (!headerFound) {
            if (isRealHeader(parsedLine)) {
                data.headers = parsedLine;
                headerFound = true;
                headerSize = parsedLine.length;
            }
            return;
        }

        while (parsedLine.length < headerSize) {
            parsedLine.push("");
        }

        if (isRowEmpty(parsedLine)) {
            return;
        }

        data.rows.push(parsedLine);
        data.sourceLineNumbers.push(lineIndex + 1);
    });

    if (!headerFound) {
        throw new Error("CSV header row was not found.");
    }

    return data;
}

// finds the column index by number or name
function findColumnIndex(headers, input) {
    const trimmedInput = input.trim();

    const number = Number(trimmedInput);
    if (!Number.isNaN(number) && Number.isInteger(number)) {
        if (number >= 1 && number <= headers.length) {
            return number - 1;
        }
    }

    for (let i = 0; i < headers.length; i++) {
        if (headers[i].trim().toLowerCase() === trimmedInput.toLowerCase()) {
            return i;
        }
    }

    return -1;
}

async function main() {
    try {
        console.log("Student: CANTAROS, RICH ANDREI");
        console.log("MP11 - Frequency Count for Column Values");

        const filePath = await askQuestion("Enter CSV dataset file path: ");
        const data = loadCSV(filePath);

        if (data.rows.length === 0) {
            console.log("No valid data rows were found in the dataset.");
            rl.close();
            return;
        }

        console.log("\nAvailable columns:");
        data.headers.forEach((header, index) => {
            if (header.trim() !== "") {
                console.log(`${index + 1} - ${header}`);
            }
        });

        const columnInput = await askQuestion("\nEnter column name or column number for frequency count: ");
        const columnIndex = findColumnIndex(data.headers, columnInput);

        if (columnIndex === -1) {
            console.log("Invalid column. Please run the program again and enter a valid column name or number.");
            rl.close();
            return;
        }

        // frequencyMap stores the count of every unique value
        const frequencyMap = {};

        data.rows.forEach(row => {
            let value = safeGet(row, columnIndex).trim();
            if (value === "") value = "(EMPTY)";

            frequencyMap[value] = (frequencyMap[value] || 0) + 1;
        });

        // sort by frequency descending, then alphabetically
        const sortedEntries = Object.entries(frequencyMap).sort((a, b) => {
            if (b[1] !== a[1]) return b[1] - a[1];
            return a[0].localeCompare(b[0]);
        });

        console.log("\n========== FREQUENCY COUNT ==========");
        console.log(`Selected Column: ${data.headers[columnIndex]}`);
        console.log(`Total Valid Rows: ${data.rows.length}`);
        console.log(`Unique Values: ${sortedEntries.length}`);
        console.log("=====================================");

        sortedEntries.forEach(([value, count]) => {
            console.log(`${value.padEnd(45)} : ${count}`);
        });

    } catch (error) {
        console.log("Error:", error.message);
    } finally {
        rl.close();
    }
}

main();